### DISCUSSION TITLE
ESP-MESH/PainlessMesh using Micropython

### DISCUSSION THREAD

**Discussion by ayenesams** [micropython/micropython#11362] [ESP32] [answered]
Working on a capstone project that uses ESPNow to communicate between sensor nodes(with ESP32s) in a spider mesh topology. 

We were recently ordered to use PainlessMesh instead. From my understanding, PainlessMesh is an Arduino library that is still pretty dependent on other Arduino libs. I can't see any Micropython support for ESP-MESH too.

Is there an existing equivalent of PainlessMesh lib/ ESP-MESH in Micropython? If not, could you give me an idea of how I'll implement this? 

tl;dr: Is there a way to port PainlessMesh onto Micropython from Arduino

**Comment by bidrohini** [comment, marked as answer]
Unfortunately, as far as I am aware, there is currently no direct equivalent of PainlessMesh or ESP-MESH in MicroPython. Both PainlessMesh and ESP-MESH are designed specifically for the ESP32 and rely on the Arduino libraries.

Porting PainlessMesh to MicroPython would be a significant undertaking, as it would require re-implementing the entire library in MicroPython. However, it may be possible to adapt some of the concepts and techniques used in PainlessMesh to work with MicroPython and the ESP32.

One option would be to implement your own mesh networking solution using the ESP32's built-in Wi-Fi and socket programming features of MicroPython.

By the way, since you are working with ESP-NOW, this project may interest you:

https://www.pcbway.com/project/shareproject/Bat_Hand_Controller_ESP32_and_Joystick_97e8ad4b.html

**Reply by villeneuve** [reply]
There is also this project: MicroPython ESP-Now https://github.com/glenn20/micropython-espnow-images

**Reply by ayenesams** [reply]
> There is also this project: MicroPython ESP-Now https://github.com/glenn20/micropython-espnow-images

Yeah I’m using these right now and I’ve already managed to make it work.

**Comment by davefes** [comment]
Still not PainlessMesh but a Mesh implementation:
https://github.com/zoland/SMesh/blob/main/smesh.py

**Reply by ayenesams** [reply]
Thanks! Will look into this.
